src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -2089,6 +2108,196 @@ SubquerySqlTaskList(Job *job, PlannerRestrictionContext *plannerRestrictionConte
+										   " the right part of right join")));
+				}
+			}
+			break;
+
+			case JOIN_FULL:
+			{
+				if (ReferenceTableExist(plannerInfo, innerrel) || ReferenceTableExist(
+						plannerInfo, outerrel))
+				{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									errmsg(
+										"can not plan query having reference table as a"
+										" part of full join")));
+				}
+			}
+			break;
+
+			default:
+			{ }
+			  break;
+		}
+	}
+}
+
+
+/*
+ * ReferenceTableExist check whether the relationInfo has reference table.
+ * Since relation ids of relationInfo indexes to the range table entry list of
+ * query, query is also passed.
+ */
+static bool
+ReferenceTableExist(PlannerInfo *plannerInfo, RelOptInfo *relationInfo)
+{
+	Relids relids = bms_copy(relationInfo->relids);
+	int relationId = -1;
+
+	while ((relationId = bms_first_member(relids)) >= 0)
+	{
+		RangeTblEntry *rangeTableEntry = plannerInfo->simple_rte_array[relationId];
+
+		/* relationInfo has this range table entry */
+		if (RTEContainsReferenceTable(rangeTableEntry))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * RTEContainsReferenceTable checks whether there exist a reference table in the
+ * given range table entry.
+ */
+bool
+RTEContainsReferenceTable(RangeTblEntry *rangeTableEntry)
+{
+	List *relationList = NIL;
+	ListCell *relationCell = NULL;
+	ExtractRTRelationFromNode((Node *) rangeTableEntry, &relationList);
+
+	foreach(relationCell, relationList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(relationCell);
+		Oid relationId = rangeTableEntry->relid;
+
+		if (IsDistributedTable(relationId) && PartitionMethod(relationId) ==
+			DISTRIBUTE_BY_NONE)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ErrorIfSetOpWithReferenceTable checks whether there exist a reference table
+ * as a part of any set operation.
+ */
+static void
+ErrorIfSetOpWithReferenceTable(Query *queryTree)
+{
+	List *joinTreeTableIndexList = NIL;
+	Index subqueryRteIndex = 0;
+	RangeTblEntry *rangeTableEntry = NULL;
+	Query *subqueryTree = NULL;
+	List *rangeTableList = queryTree->rtable;
+	Node *setOperations = queryTree->setOperations;
+	ExtractRangeTableIndexWalker((Node *) queryTree->jointree, &joinTreeTableIndexList);
+
+	if (setOperations != NULL)
+	{
+		List *rangeTableList = NIL;
+		ListCell *rangeTableCell = NULL;
+		ExtractRangeTableRelationWalker((Node *) queryTree, &rangeTableList);
+
+		foreach(rangeTableCell, rangeTableList)
+		{
+			RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+			Oid relationId = rangeTableEntry->relid;
+			if (PartitionMethod(relationId) == DISTRIBUTE_BY_NONE)
+			{
+				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								errmsg(
+									"can not plan query having reference table with union")));
+			}
+		}
+	}
+
+	if (list_length(joinTreeTableIndexList) < 1)
+	{
+		return;
+	}
+
+	subqueryRteIndex = linitial_int(joinTreeTableIndexList);
```

Comment:
why do we want to limit this with the first subquery only ? Would it be fine if we have set operation in the second subquery participated into join ?